### PR TITLE
Upgrade ReactiveMongo to 0.18.3

### PR DESCRIPTION
**Pull request by cchantep**

### CODE REVIEW

File: common/src/main/scala/akka/contrib/persistence/mongodb/MongoCollectionCache.scala [hunk 1]
```diff
@@ -48,8 +48,10 @@ object MongoCollectionCache {
     override def getOrElseCreate(collectionName: String, collectionCreator: String => C): C =
       trieMap.getOrElseUpdate(collectionName, collectionCreator.apply(collectionName))
 
-    override def invalidate(collectionName: String): Unit =
+    override def invalidate(collectionName: String): Unit = {
       trieMap.remove(collectionName)
+      ()
```
Comment (cchantep) [inline common/src/main/scala/akka/contrib/persistence/mongodb/MongoCollectionCache.scala]: Discarded value as `Unit`

File: common/src/main/scala/akka/contrib/persistence/mongodb/MongoCollectionCache.scala [hunk 2]
```diff
@@ -76,8 +78,10 @@ object MongoCollectionCache {
       }
     }
 
-    override def invalidate(collectionName: String): Unit =
+    override def invalidate(collectionName: String): Unit = {
       trieMap.remove(collectionName)
+      ()
```
Comment (cchantep) [inline common/src/main/scala/akka/contrib/persistence/mongodb/MongoCollectionCache.scala]: Discarded value as `Unit`

File: rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoJournaller.scala [hunk 3]
```diff
@@ -81,12 +81,17 @@ class RxMongoJournaller(val driver: RxMongoDriver) extends MongoPersistenceJourn
   private[this] def doBatchAppend(batch: Seq[Try[BSONDocument]], collection: Future[BSONCollection])(implicit ec: ExecutionContext): Future[Seq[Try[BSONDocument]]] = {
     if (batch.forall(_.isSuccess)) {
       val collected = batch.toStream.collect { case Success(doc) => doc }
-      collection.flatMap(_.insert[BSONDocument](ordered = true, writeConcern).many(collected).map(_ => batch))
+
```
Comment (cchantep) [inline rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoJournaller.scala]: Refactoring

File: rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoJournaller.scala [hunk 4]
```diff
@@ -134,10 +139,7 @@ class RxMongoJournaller(val driver: RxMongoDriver) extends MongoPersistenceJourn
         Match(BSONDocument(PROCESSOR_ID -> persistenceId, TO -> BSONDocument("$lte" -> maxSequenceNr))),
         GroupField(PROCESSOR_ID)("max" -> MaxField(TO)) :: Nil,
         batchSize = Option(1)
-      ).prepared[AkkaStreamCursor]
-        .cursor
-        .headOption
-        .map(_.flatMap(_.getAs[Long]("max")))
+      ).prepared.cursor.headOption.map(_.flatMap(_.getAs[Long]("max")))
```
Comment (cchantep) [inline rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoJournaller.scala]: Not specific to `AkkaStreamCursor`

File: rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoPersistenceExtension.scala [hunk 5]
```diff
@@ -9,12 +9,18 @@ package akka.contrib.persistence.mongodb
...
+  DefaultDB,
+  FailoverStrategy,
+  MongoConnection,
+  MongoDriver
+}
 import reactivemongo.api.collections.bson.BSONCollection
-import reactivemongo.api.commands._
+import reactivemongo.api.commands.{CommandError, WriteConcern}
 import reactivemongo.api.indexes.{Index, IndexType}
-import reactivemongo.bson._
```
Comment (cchantep) [inline rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoPersistenceExtension.scala]: Avoid ambiguity/conflict due to wildcard imports

File: rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoPersistenceExtension.scala [hunk 6]
```diff
@@ -133,21 +139,10 @@ class RxMongoDriver(system: ActorSystem, config: Config, driverProvider: RxMongo
       _ <- if (s.capped) Future.successful(()) else cc.convertToCapped(realtimeCollectionSize, None)
     } yield cc
 
-  private[mongodb] def getCollections(collectionName: String)(implicit ec: ExecutionContext): Enumerator[BSONCollection] = {
```
Comment (cchantep) [inline rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoPersistenceExtension.scala]: Unused (avoid dependency with iteratees)

File: rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoReadJournaller.scala [hunk 7]
```diff
@@ -143,15 +142,15 @@ object CurrentEventsByTag {
   }
 }
 
-class RxMongoRealtimeGraphStage(driver: RxMongoDriver, bufsz: Int = 16)(factory: Option[BSONObjectID] => Publisher[BSONDocument])
```
Comment (cchantep) [inline rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoReadJournaller.scala]: Unused, avoid later conversion of `bufsz`

File: rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoReadJournaller.scala [hunk 8]
```diff
@@ -199,21 +198,21 @@ class RxMongoRealtimeGraphStage(driver: RxMongoDriver, bufsz: Int = 16)(factory:
         override def onComplete(): Unit = cmpAcImpl.invoke(())
       }
 
-      setHandler(out, new OutHandler {
```
Comment (cchantep) [inline rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoReadJournaller.scala]: Refactoring (avoid double instantiation)

File: rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoSnapshotter.scala [hunk 9]
```diff
@@ -44,9 +44,9 @@ class RxMongoSnapshotter(driver: RxMongoDriver) extends MongoPersistenceSnapshot
 
     for {
       s <- snaps(pid)
-      wr <- s.remove(BSONDocument(criteria: _*), writeConcern)
+      wr <- s.delete(true, writeConcern).one(BSONDocument(criteria: _*))
     } yield {
-      if (driver.useSuffixedCollectionNames && driver.suffixDropEmpty && wr.ok)
```
Comment (cchantep) [inline rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoSnapshotter.scala]: Write operation never return a `Future.successful[WriteResult]` if not ok (and as already in `yield` ...)

**Comment by scullxbones** [comment]
Thanks for the PR!

I've restarted the travis suite, but it seems there's a drop-off in stability with these changes.  There are a few race conditions left in the suite, so it doesn't fully pass, but a majority of the scala/jvm/mongo tuples failed.

I'm not seeing anything jump out at me w.r.t. the PR contents.

Will keep an eye on it.

**Comment by cchantep** [comment]
Except Travis/Docker instability, I can hardly see how the failures are related to the changes.

**Comment by scullxbones** [comment]
I agree, I don't think the failures are the result of the code changes in the project.

During the lifetime of this project, some versions of the reactivemongo driver have been more stable under test than others.  For example version `0.16` was quite stable, most builds would fully succeed for the matrix of scala 2.11, 2.12 x mongo 3.0, 3.2, 3.4, 3.6.  Occasionally one or two of the eight would fail.

If you see anything obviously wrong in how the driver is used during the test suite to eliminate the instability in the face of upgrades that would be very helpful.

**Comment by cchantep** [comment]
The driver versions 0.16.6+ (not 0.16.0+), including 0.17.x and 0.18.x, are on their side more stable according integration tests & benchmarks, as some pool/nodeset refactoring has been achieved (channel dedidated to signaling, configurable max-in-flight requests, connection load-balancing, ...).

**Comment by scullxbones** [comment]
I pulled this branch locally to see if I could duplicate what travis sees.  The tests are consistently failing locally for me.  The `AuthSpec` seems especially problematic.  Are they not failing for you?

A minimal set can be run with: `sbt akka-persistence-mongo-rxmongo/travis:test`.  Test docker containers can be created with `./test_containers.sh`

The test suite was much more stable with `0.16.0` with rare false positives.  I tried a few driver versions including `0.16.6`, `0.17.1`, and `0.18.4` and all are not stable.  

Again, this could be an issue with the test suite, so if you see anything obvious let me know.  I'd much rather have stable tests and be able to upgrade the reactivemongo driver version.

**Comment by scullxbones** [comment]
Another couple data points, I tested 2 cases:

- this branch but with version `0.16.1` and minimal changes  ~> result was stable
- master branch with version `0.16.1` was just the `build.sbt` change ~> result was stable

I also tested a minimal compiling version with the `0.18.4` driver upgrade and it was also unstable, similar with the `AuthSpec` failing.

So i'm confident that the changes themselves are not causing the instability

**Pull request closed without merging**